src/extension/infer.rs

```diff
@@ -663,31 +680,85 @@ impl UnificationContext {
+            scope_size = new_scope_size;
+            for m in variable_scope.clone().iter() {
+                for c in self.get_constraints(m).unwrap().iter() {
+                    match c {
+                        Constraint::Plus(_, other) => variable_scope.insert(self.resolve(*other)),
+                        Constraint::Equal(other) => variable_scope.insert(self.resolve(*other)),
+                    };
+                }
+            }
+            new_scope_size = variable_scope.len();
+        }
+        variable_scope
+    }
+
     /// Instantiate all variables in the graph with the empty extension set, or
     /// the smallest solution possible given their constraints.
     /// This is done to solve metas which depend on variables, which allows
     /// us to come up with a fully concrete solution to pass into validation.
+    ///
+    /// Nodes which loop into themselves must be considered as a "minimum" set
+    /// of requirements. If we have
+    ///   1 = 2 + X, ...
+    ///   2 = 1 + x, ...
+    /// then 1 and 2 both definitely contain X, even if we don't know what else.
+    /// So instead of instantiating to the empty set, we'll instantiate to `{X}`
     pub fn instantiate_variables(&mut self) {
-        for m in self.variables.clone().into_iter() {
+        // A directed graph to keep track of `Plus` constraint relationships
+        let mut relations = GraphContainer::new_directed();
+        let mut solutions: HashMap<Meta, ExtensionSet> = HashMap::new();
+
+        let variable_scope = self.search_variable_deps();
+        for m in variable_scope.into_iter() {
             if !self.solved.contains_key(&m) {
                 // Handle the case where the constraints for `m` contain a self
                 // reference, i.e. "m = Plus(E, m)", in which case the variable
                 // should be instantiated to E rather than the empty set.
-                let solution =
-                    ExtensionSet::from_iter(self.get_constraints(&m).unwrap().iter().filter_map(
-                        |c| match c {
-                            // If `m` has been merged, [`self.variables`] entry
-                            // will have already been updated to the merged
-                            // value by [`self.merge_equal_metas`] so we don't
-                            // need to worry about resolving it.
-                            Constraint::Plus(x, other_m) if m == self.resolve(*other_m) => {
-                                Some(x.clone())
-                            }
-                            _ => None,
-                        },
-                    ));
-                self.add_solution(m, solution);
+                let mut solution = ExtensionSet::new();
+                // Is the solution complete - does this variable *only* refer to itself?
+                let mut solution_complete = true;
+
+                for c in self.get_constraints(&m).unwrap().iter() {
```

Comment:
Trying a more-functional approach. See what you think.
```suggestion
let plus_constraints = self.get_constraints(&m).unwrap().iter().filter_map(match |c| {
   Constraint::Plus(r, other_m) => Some((r, self.resolve(*other_m)))
   _ => None
})
let (rs, other_ms) : (Vec<_>, Vec<_>) = plus_constraints.unzip();
let solution = ExtensionSet::from(rs); // We must have something like that right?
let unresolved_ms = other_ms.filter(|other_m| m != other_m); // probably other ways to do that!
if unresolved_ms.empty() {
    self.add_solution(m, solution);
} else {
    unresolved_ms.for_each(|other_m| relations.add_edge(m,other_m));
}
solutions.insert(m, solution);
```

Reply:
Nice!